### PR TITLE
add support for wildcard, inversion tacticals

### PR DESCRIPTION
**Pull request by jonsterling**
This is a nice tactic. For instance, consider the following code from `torus.red`:

```
let c2t2c/transpose : (c0 c1 : s1) → path (s1 × s1) (t2c (c2t/transpose c0 c1)) (c0, c1) =
  λ [ _ → λ [ _ → refl ] ]

```

The declaration of `c2t2c/transpose` gets elaborated to the following monster:

```
(λ [x]
 (s1.elim
  [x1] (→ [c1 : s1]
          (# <i>
              (× [x1 : s1] s1)
              [i=0
               (torus.elim
                [x1] (× [x1 : s1] s1)
                ((s1.elim
                  [x1] (→ [x1 : s1] torus)
                  x1
                  (base (λ [x1] (s1.elim [x1] torus x1 (base pt) (loop [x1] (p/two x1)))))
                  (loop
                   [x1]
                    (λ [x1]
                     (s1.elim [x1] torus x1 (base (p/one x1)) (loop [x1] (square x1 x1))))))
                 c1)
                (pt (pair base base))
                (p/one [x1] (pair (loop x1) base))
                (p/two [x1] (pair base (loop x1)))
                (square [x1 x1] (pair (loop x1) (loop x1))))]
              [i=1 (pair x1 c1)]))
  x
  (loop
   [i]
    (λ [x1]
     (s1.elim
      [x1] (# <i1>
               (× [x1 : s1] s1)
               [i1=0
                (torus.elim
                 [x1] (× [x1 : s1] s1)
                 (s1.elim [x1] torus x1 (base (p/one i)) (loop [x1] (square x1 i)))
                 (pt (pair base base))
                 (p/one [x1] (pair (loop x1) base))
                 (p/two [x1] (pair base (loop x1)))
                 (square [x1 x1] (pair (loop x1) (loop x1))))]
               [i1=1 (pair (loop i) x1)])
      x1
      (loop [i1] (λ <i1> (pair (loop i) (loop i1))))
      (base (λ <i1> (pair (loop i) base))))))
  (base
   (λ [x1]
    (s1.elim
     [x1] (# <i>
              (× [x1 : s1] s1)
              [i=0
               (torus.elim
                [x1] (× [x1 : s1] s1)
                (s1.elim [x1] torus x1 (base pt) (loop [x1] (p/two x1)))
                (pt (pair base base))
                (p/one [x1] (pair (loop x1) base))
                (p/two [x1] (pair base (loop x1)))
                (square [x1 x1] (pair (loop x1) (loop x1))))]
              [i=1 (pair base x1)])
     x1
     (loop [i] (λ <i1> (pair base (loop i))))
     (base (λ <i> (pair base base))))))))
```

### CODE REVIEW

**Comment by favonia** [comment]
Will it be unclear to a user that `λ [ _ → blah ]` and `λ _ → blah` are completely different?

**Comment by jonsterling** [comment]
@favonia I'm not sure :) Actually what I really wanted was "[* -> foo]", but couldn't do it for lexing reasons.

**Comment by favonia** [comment]
> but couldn't do it for lexing reasons.

Is it some conflict in the LR(1) grammar or what?

**Comment by jonsterling** [comment]
@favonia It's just that \times and * get lexed as the same symbol.

**Comment by jonsterling** [comment]
I guess it's easy enough to fix :laughing:

File: example/or.red [hunk 1]
```diff
@@ -1,24 +1,12 @@
...
-  (C : type)
-  (t : or A B)
-  (m0 : A → C)
-  (m1 : B → C)
-  : C =
-  sg/elim bool _ (λ _ _ → C) t
-    (λ b →
-      elim b [tt → m0 | ff → m1])
+let or/elim (A B : type) (C : type) : (t : or A B) (m0 : A → C) (m1 : B → C) → C =
+  λ [,] →
```
Comment (jonsterling) [inline example/or.red]: This gadget takes a goal `(A * B) -> C` to `A -> B -> C`. Makes it easier to work with sigma types on the left of the sequent!

File: example/torus.red [hunk 2]
```diff
@@ -15,58 +15,40 @@ data torus where
...
-  λ [
+  elim [
   | base → pt
   | loop i → p/two i
   ]
 
-let c2t/transpose : s1 → s1 → torus =
-  λ [
+let c2t : (s1 × s1) → torus =
+  λ [,] →
```
Comment (jonsterling) [inline example/torus.red]: Using this inversion gadget, we avoid the two-step definition that we had before.

File: example/s3-to-join.red [hunk 3]
```diff
@@ -67,9 +67,8 @@ let join→s3/push (a b : s1) : path s3 base base =
 
 let join→s3 : join → s3 =
   elim [
-  | inl a → base
-  | inr b → base
   | push a b i → join→s3/push a b i
+  | * → base
```
[the remark this answers is not in the document]
Comment (jonsterling) [reply example/s3-to-join.red]: Taste it!

**Pull request merged**